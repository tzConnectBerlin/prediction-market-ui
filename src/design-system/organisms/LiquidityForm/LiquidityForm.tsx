import * as React from 'react';
import * as Yup from 'yup';
import { Field, Form, Formik, FormikHelpers, FormikProps } from 'formik';
import { useTranslation } from 'react-i18next';
import { Grid } from '@material-ui/core';
import { FormikTextField } from '../../molecules/FormikTextField';
import { CustomButton } from '../../atoms/Button';
import { Typography } from '../../atoms/Typography';
import { MarketTradeType, Token, TokenType } from '../../../interfaces';
import { PositionItem, PositionSummary } from '../SubmitBidCard/PositionSummary';
import { getNoTokenId, getTokenQuantityById, getYesTokenId } from '../../../utils/misc';
import {
  liquidityTokensMovedToPool,
  poolShareCalculation,
  tokensMovedToPool,
} from '../../../contracts/MarketCalculations';
import { roundToTwo, tokenDivideDown, tokenMultiplyUp } from '../../../utils/math';

const TokenPriceDefault = {
  yes: 0,
  no: 0,
};
export type LiquidityValue = {
  yesToken: string | number;
  noToken: string | number;
  tradeType: MarketTradeType;
};
export interface LiquidityFormProps {
  /**
   * Callback to get the form values
   */
  handleSubmit: (
    values: LiquidityValue,
    formikHelpers: FormikHelpers<LiquidityValue>,
  ) => void | Promise<void>;
  /**
   * Initial values to use when initializing the form. Default is 0.
   */
  initialValues?: Omit<LiquidityValue, 'tradeType'>;
  /**
   * Market Id
   */
  marketId: string;
  /**
   * Pool token values
   */
  poolTokens?: Token[];
  /**
   * Pool token Amount
   */
  poolTotalSupply: number;
  /**
   * User token values
   */
  userTokens?: Token[];
  /**
   * Title form to display
   */
  title: string;
  /**
   * TokenName to display
   */
  tokenName?: string;
  /**
   * Liquidity TokenName to display
   */
  liquidityTokenName?: string;
  /**
   * Trade type
   */
  tradeType: MarketTradeType;
  /**
   * Is wallet connected
   */
  connected?: boolean;
  /**
   * Token Price
   */
  tokenPrice?: {
    yes: number;
    no: number;
  };
}

export const LiquidityForm: React.FC<LiquidityFormProps> = ({
  title,
  tokenName = 'PMM',
  liquidityTokenName = 'LQT',
  handleSubmit,
  connected,
  tradeType,
  poolTokens,
  userTokens,
  marketId,
  poolTotalSupply,
  tokenPrice = TokenPriceDefault,
}) => {
  const { t } = useTranslation('common');
  const yesTokenId = React.useMemo(() => getYesTokenId(marketId), [marketId]);
  const noTokenId = React.useMemo(() => getNoTokenId(marketId), [marketId]);
  const [pools, setPools] = React.useState({
    yesPool: 0,
    noPool: 0,
  });
  const [userAmounts, setUserAmounts] = React.useState({
    yesToken: 0,
    noToken: 0,
  });
  const [poolShare, setPoolShare] = React.useState(0);
  const [expectedBalance, setExpectedBalance] = React.useState<PositionItem[]>([]);
  const [expectedStake, setExpectedStake] = React.useState<PositionItem[]>([]);
  const [adjustedBalance, setAdjustedBalance] = React.useState<PositionItem[]>([]);
  const [adjustedStake, setAdjustedStake] = React.useState<PositionItem[]>([]);

  React.useEffect(() => {
    if (poolTokens) {
      const yesPool = getTokenQuantityById(poolTokens, yesTokenId);
      const noPool = getTokenQuantityById(poolTokens, noTokenId);
      setPools({
        yesPool,
        noPool,
      });
    }
    if (userTokens) {
      const yesToken = getTokenQuantityById(userTokens, yesTokenId);
      const noToken = getTokenQuantityById(userTokens, noTokenId);
      setUserAmounts({
        noToken,
        yesToken,
      });
    }
  }, [poolTotalSupply, poolTokens, userTokens, yesTokenId, noTokenId]);

  const validationSchema = Yup.object({
    yesToken: Yup.number().min(0.000001, `Min quantity is 0.000001`).required('Required'),
    noToken: Yup.number().min(0.000001, `Min quantity is 0.000001`).required('Required'),
  });

  const initialFormValues: LiquidityValue = {
    yesToken: '',
    noToken: '',
    tradeType,
  };

  const handleChange = (e: any, setFieldValue: any, fieldName: string, tokenType: TokenType) => {
    if (connected) {
      const [aPool, bPool] =
        TokenType.yes === tokenType ? [pools.yesPool, pools.noPool] : [pools.noPool, pools.yesPool];
      const aToken = tokenMultiplyUp(e.target.value);
      const l = liquidityTokensMovedToPool(aToken, aPool, poolTotalSupply);
      setPoolShare(poolShareCalculation(l, poolTotalSupply));
      const bToken = tokensMovedToPool(bPool, poolShare);
      setFieldValue(fieldName, roundToTwo(tokenDivideDown(bToken)));
      const [newYes, newNo] = TokenType.yes === tokenType ? [aToken, bToken] : [bToken, aToken];
      const expectedValue = tokenPrice.yes * newYes + tokenPrice.no * newNo;
      const expectedTotalValue =
        tokenPrice.yes * (userAmounts.yesToken + newYes) +
        tokenPrice.no * (userAmounts.noToken + newNo);
      const newExpextedStake: PositionItem[] = [
        {
          label: t('liquidityTokens'),
          value: `${roundToTwo(tokenDivideDown(l))}${liquidityTokenName}`,
        },
        {
          label: t('stakeInPool'),
          value: `${roundToTwo(tokenDivideDown(poolShare))}%`,
        },
        {
          label: t('value'),
          value: `${roundToTwo(tokenDivideDown(expectedValue))}${tokenName}`,
        },
      ];
      setExpectedStake(newExpextedStake);
      const newExpextedBalance: PositionItem[] = [
        {
          label: t('yesTokens'),
          value: `${roundToTwo(tokenDivideDown(userAmounts.yesToken + newYes))}(-${roundToTwo(
            tokenDivideDown(newYes),
          )})`,
        },
        {
          label: t('noTokens'),
          value: `${roundToTwo(tokenDivideDown(userAmounts.noToken + newNo))}(-${roundToTwo(
            tokenDivideDown(newNo),
          )})`,
        },
        {
          label: t('value'),
          value: `${roundToTwo(tokenDivideDown(expectedTotalValue))}${tokenName}`,
        },
      ];
      setExpectedBalance(newExpextedBalance);
    }
  };
  return (
    <>
      {tradeType === MarketTradeType.payIn && (
        <Grid container direction="column" spacing={2}>
          <Grid item>
            <Formik
              onSubmit={handleSubmit}
              validationSchema={validationSchema}
              initialValues={initialFormValues}
              enableReinitialize
            >
              {({ isValid, setFieldValue }) => (
                <Form>
                  <Grid
                    container
                    spacing={3}
                    direction="column"
                    alignContent="flex-start"
                    justifyContent="center"
                  >
                    <Grid item container direction="column" width="100%">
                      <Grid item>
                        <Field
                          component={FormikTextField}
                          label={t('amount')}
                          name="yesTokens"
                          type="number"
                          pattern="[0-9]*"
                          placeholder="Type here"
                          handleChange={(e: any) =>
                            handleChange(e, setFieldValue, 'noTokens', TokenType.yes)
                          }
                          fullWidth
                          InputProps={{
                            endAdornment: (
                              <Typography color="text.secondary" component="span">
                                {t('yesTokens')}
                              </Typography>
                            ),
                          }}
                          required
                        />
                      </Grid>
                      <Grid item>
                        <Field
                          component={FormikTextField}
                          label=""
                          name="noTokens"
                          type="number"
                          pattern="[0-9]*"
                          placeholder="Type here"
                          handleChange={(e: any) =>
                            handleChange(e, setFieldValue, 'yesTokens', TokenType.no)
                          }
                          fullWidth
                          InputProps={{
                            endAdornment: (
                              <Typography color="text.secondary" component="span">
                                {t('noTokens')}
                              </Typography>
                            ),
                          }}
                        />
                      </Grid>
                    </Grid>
                    {connected && (
                      <>
                        {expectedStake.length > 0 && (
                          <Grid item>
                            <PositionSummary title={t('expectedStake')} items={expectedStake} />
                          </Grid>
                        )}
                        {expectedBalance.length > 0 && (
                          <Grid item>
                            <PositionSummary title={t('expectedBalance')} items={expectedBalance} />
                          </Grid>
                        )}
                        {adjustedStake.length > 0 && (
                          <Grid item>
                            <PositionSummary title={t('adjustedStake')} items={adjustedStake} />
                          </Grid>
                        )}
                        {adjustedBalance.length > 0 && (
                          <Grid item>
                            <PositionSummary title={t('adjustedBalance')} items={adjustedBalance} />
                          </Grid>
                        )}
                      </>
                    )}
                    <Grid item flexDirection="column">
                      <CustomButton
                        color="primary"
                        type="submit"
                        label={!connected ? `${t('connectWallet')} + ${t(title)}` : t(title)}
                        fullWidth
                        disabled={!isValid}
                      />
                      <Typography size="body1" mt="1rem">
                        {t('requiredField')}
                      </Typography>
                    </Grid>
                  </Grid>
                </Form>
              )}
            </Formik>
          </Grid>
        </Grid>
      )}
      {/* {!connected && tradeType === MarketTradeType.payOut && (
        <Typography size="body2">Only liquidity providers can remove liquidity</Typography>
      )} */}
    </>
  );
};
