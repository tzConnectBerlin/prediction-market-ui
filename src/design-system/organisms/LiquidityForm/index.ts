export { LiquidityForm, LiquidityForm as default } from './LiquidityForm';
export type { LiquidityFormProps } from './LiquidityForm';
