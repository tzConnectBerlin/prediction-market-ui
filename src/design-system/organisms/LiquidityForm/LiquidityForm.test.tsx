import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { QueryClientProvider, QueryClient } from 'react-query';
import { ThemeProvider } from '@mui/system';
import { LiquidityForm, LiquidityFormProps } from './LiquidityForm';
import { lightTheme } from '../../../styles/theme';

const basicArgs = {
  tokenName: 'PMM',
  title: 'Add Liquidity',
  marketId: '3',
  poolTotalSupply: 100,
};
const queryClient = new QueryClient();
const defaultArgs: LiquidityFormProps = {
  ...basicArgs,
  operationType: 'add',
  handleSubmit: jest.fn(),
};

const WrappedComponent = (props: LiquidityFormProps) => (
  <QueryClientProvider client={queryClient}>
    <ThemeProvider theme={lightTheme}>
      <LiquidityForm {...props} />
    </ThemeProvider>
  </QueryClientProvider>
);
describe('Snapshot - render LiquidityForm', () => {
  it('renders correctly with default props', () => {
    const AddLiquidityForm = renderer.create(<WrappedComponent {...defaultArgs} />).toJSON();
    expect(AddLiquidityForm).toMatchSnapshot();
  });
});

describe('Element testing LiquidityForm Component', () => {
  it('render correctly LiquidityForm with default props', async () => {
    const { getAllByText } = render(<WrappedComponent {...defaultArgs} connected />);
    expect(getAllByText(/ Liquidity/i).length).toBe(1);
  });
});
