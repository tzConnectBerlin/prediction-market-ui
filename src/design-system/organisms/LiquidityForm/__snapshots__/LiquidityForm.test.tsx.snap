// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot - render LiquidityForm renders correctly with default props 1`] = `
.emotion-0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-top: -16px;
  width: calc(100% + 16px);
  margin-left: -16px;
}

.emotion-0>.MuiGrid-item {
  max-width: none;
}

.emotion-0>.MuiGrid-item {
  padding-top: 16px;
}

.emotion-0>.MuiGrid-item {
  padding-left: 16px;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-2 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-top: -24px;
  width: calc(100% + 24px);
  margin-left: -24px;
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: flex-start;
  align-content: flex-start;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-2>.MuiGrid-item {
  max-width: none;
}

.emotion-2>.MuiGrid-item {
  padding-top: 24px;
}

.emotion-2>.MuiGrid-item {
  padding-left: 24px;
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  width: 100%;
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  min-width: 0;
  padding: 0;
  margin: 0;
  border: 0;
  vertical-align: top;
}

.emotion-5 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-6 {
  box-sizing: border-box;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-basis: 91.666667%;
  -ms-flex-preferred-size: 91.666667%;
  flex-basis: 91.666667%;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  max-width: 91.666667%;
}

.emotion-7 {
  color: rgba(0, 0, 0, 0.6);
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.4375em;
  letter-spacing: 0.00938em;
  padding: 0;
  position: relative;
  display: block;
  transform-origin: top left;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 133%;
  position: absolute;
  left: 0;
  top: 0;
  -webkit-transform: translate(0, -1.5px) scale(0.75);
  -moz-transform: translate(0, -1.5px) scale(0.75);
  -ms-transform: translate(0, -1.5px) scale(0.75);
  transform: translate(0, -1.5px) scale(0.75);
  -webkit-transition: color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,-webkit-transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,max-width 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,max-width 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-weight: bold;
  color: #1976d2;
  position: relative;
  -webkit-transform: none;
  -moz-transform: none;
  -ms-transform: none;
  transform: none;
  -webkit-transition: none;
  transition: none;
  font-size: 0.8rem;
}

.emotion-7.Mui-focused {
  color: #1976d2;
}

.emotion-7.Mui-disabled {
  color: rgba(0, 0, 0, 0.38);
}

.emotion-7.Mui-error {
  color: #d32f2f;
}

.emotion-7.Mui-disabled {
  color: #1976d2;
  opacity: 0.38;
}

.emotion-7 .label-asterisk {
  line-height: 0;
}

.emotion-8 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.75;
  letter-spacing: 0.00938em;
}

.emotion-8.truncate {
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 3;
}

.emotion-9.Mui-error {
  color: #d32f2f;
}

.emotion-10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  min-width: 0;
  padding: 0;
  margin: 0;
  border: 0;
  vertical-align: top;
  width: 100%;
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
}

.emotion-10 input {
  text-align: inherit;
}

.emotion-11 {
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.4375em;
  letter-spacing: 0.00938em;
  color: rgba(0, 0, 0, 0.87);
  box-sizing: border-box;
  position: relative;
  cursor: text;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  position: relative;
}

.emotion-11.Mui-disabled {
  color: rgba(0, 0, 0, 0.38);
  cursor: default;
}

label+.emotion-11 {
  margin-top: 16px;
}

.emotion-11:after {
  border-bottom: 2px solid #1976d2;
  left: 0;
  bottom: 0;
  content: "";
  position: absolute;
  right: 0;
  -webkit-transform: scaleX(0);
  -moz-transform: scaleX(0);
  -ms-transform: scaleX(0);
  transform: scaleX(0);
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  pointer-events: none;
}

.emotion-11.Mui-focused:after {
  -webkit-transform: scaleX(1);
  -moz-transform: scaleX(1);
  -ms-transform: scaleX(1);
  transform: scaleX(1);
}

.emotion-11.Mui-error:after {
  border-bottom-color: #d32f2f;
  -webkit-transform: scaleX(1);
  -moz-transform: scaleX(1);
  -ms-transform: scaleX(1);
  transform: scaleX(1);
}

.emotion-11:before {
  border-bottom: 1px solid rgba(0, 0, 0, 0.42);
  left: 0;
  bottom: 0;
  content: "\\00a0";
  position: absolute;
  right: 0;
  -webkit-transition: border-bottom-color 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: border-bottom-color 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  pointer-events: none;
}

.emotion-11:hover:not(.Mui-disabled):before {
  border-bottom: 2px solid rgba(0, 0, 0, 0.87);
}

@media (hover: none) {
  .emotion-11:hover:not(.Mui-disabled):before {
    border-bottom: 1px solid rgba(0, 0, 0, 0.42);
  }
}

.emotion-11.Mui-disabled:before {
  border-bottom-style: dotted;
}

.emotion-12 {
  font: inherit;
  letter-spacing: inherit;
  color: currentColor;
  padding: 4px 0 5px;
  border: 0;
  box-sizing: content-box;
  background: none;
  height: 1.4375em;
  margin: 0;
  -webkit-tap-highlight-color: transparent;
  display: block;
  min-width: 0;
  width: 100%;
  -webkit-animation-name: mui-auto-fill-cancel;
  animation-name: mui-auto-fill-cancel;
  -webkit-animation-duration: 10ms;
  animation-duration: 10ms;
}

.emotion-12::-webkit-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-12::-moz-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-12:-ms-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-12::-ms-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-12:focus {
  outline: 0;
}

.emotion-12:invalid {
  box-shadow: none;
}

.emotion-12::-webkit-search-decoration {
  -webkit-appearance: none;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-12::-webkit-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-12::-moz-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-12:-ms-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-12::-ms-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-12:focus::-webkit-input-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-12:focus::-moz-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-12:focus:-ms-input-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-12:focus::-ms-input-placeholder {
  opacity: 0.42;
}

.emotion-12.Mui-disabled {
  opacity: 1;
  -webkit-text-fill-color: rgba(0, 0, 0, 0.38);
}

.emotion-12:-webkit-autofill {
  -webkit-animation-duration: 5000s;
  animation-duration: 5000s;
  -webkit-animation-name: mui-auto-fill;
  animation-name: mui-auto-fill;
}

.emotion-13 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
  color: rgba(0, 0, 0, 0.6);
  white-space: nowrap;
}

.emotion-13.truncate {
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 3;
}

.emotion-23 {
  box-sizing: border-box;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-top: 0.5rem;
}

.emotion-24 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.8125rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  min-width: 64px;
  padding: 4px 10px;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: #fff;
  background-color: #1976d2;
  box-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2),0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);
  width: 100%;
  border-radius: 0.2em;
  padding: 0.2em 1.2em;
  border: solid 2px transparent;
  text-transform: none;
  box-shadow: none;
}

.emotion-24::-moz-focus-inner {
  border-style: none;
}

.emotion-24.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-24 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-24:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #1565c0;
  box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);
}

@media (hover: none) {
  .emotion-24:hover {
    background-color: #1976d2;
  }
}

.emotion-24:active {
  box-shadow: 0px 5px 5px -3px rgba(0,0,0,0.2),0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);
}

.emotion-24.Mui-focusVisible {
  box-shadow: 0px 3px 5px -1px rgba(0,0,0,0.2),0px 6px 10px 0px rgba(0,0,0,0.14),0px 1px 18px 0px rgba(0,0,0,0.12);
}

.emotion-24.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
  box-shadow: none;
  background-color: rgba(0, 0, 0, 0.12);
}

.emotion-24:hover {
  border-width: 2px;
  box-shadow: none;
}

.emotion-24:disabled {
  border-color: transparent;
}

.emotion-25 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 3rem;
  line-height: 1.167;
  letter-spacing: 0em;
}

.emotion-25.truncate {
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 3;
}

.emotion-26 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
  margin-top: 1rem;
}

.emotion-26.truncate {
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 3;
}

<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column emotion-0"
>
  <div
    className="MuiGrid-root MuiGrid-item emotion-1"
  >
    <form
      action="#"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-direction-xs-column emotion-2"
      >
        <div
          className="MuiGrid-root MuiGrid-item emotion-3"
        >
          <div
            className="MuiFormControl-root emotion-4"
          >
            <div
              className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined emotion-5"
            >
              <div
                className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-11 emotion-6"
              >
                <label
                  className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required emotion-7"
                  data-shrink={true}
                  htmlFor="yesToken"
                >
                  <span
                    className="MuiTypography-root MuiTypography-subtitle1  emotion-8"
                    lines={3}
                  >
                    amount
                  </span>
                  <span
                    aria-hidden={true}
                    className="label-asterisk MuiInputLabel-asterisk MuiFormLabel-asterisk emotion-9"
                  >
                     
                    *
                  </span>
                </label>
              </div>
            </div>
            <div
              alignright={false}
              className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root emotion-10"
              pattern="[0-9]*"
            >
              <div
                className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd emotion-11"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd emotion-12"
                  disabled={false}
                  name="yesToken"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="inputFieldPlaceholder"
                  required={false}
                  type="number"
                  value=""
                />
                <span
                  className="MuiTypography-root MuiTypography-body1  emotion-13"
                  lines={3}
                >
                  yesTokens
                </span>
              </div>
            </div>
            
          </div>
          <div
            className="MuiFormControl-root emotion-4"
          >
            <div
              className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined emotion-5"
            >
              <div
                className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-11 emotion-6"
              >
                <label
                  className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary emotion-7"
                  data-shrink={true}
                  htmlFor="noToken"
                >
                  <span
                    className="MuiTypography-root MuiTypography-subtitle1  emotion-8"
                    lines={3}
                  >
                    
                  </span>
                </label>
              </div>
            </div>
            <div
              alignright={false}
              className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root emotion-10"
              pattern="[0-9]*"
            >
              <div
                className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd emotion-11"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd emotion-12"
                  disabled={false}
                  name="noToken"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="inputFieldPlaceholder"
                  required={false}
                  type="number"
                  value=""
                />
                <span
                  className="MuiTypography-root MuiTypography-body1  emotion-13"
                  lines={3}
                >
                  noTokens
                </span>
              </div>
            </div>
            
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item emotion-23"
        >
          <button
            borderColor="transparent"
            className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-fullWidth MuiButtonBase-root emotion-24"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            textType="none"
            type="submit"
          >
            <h3
              className="MuiTypography-root MuiTypography-h3  emotion-25"
              lines={3}
            >
              connectWalletContinue
            </h3>
          </button>
          <p
            className="MuiTypography-root MuiTypography-body1  emotion-26"
            lines={3}
          >
            requiredField
          </p>
        </div>
      </div>
    </form>
  </div>
</div>
`;
